lib/internal/asyncEachOfLimit.js

```diff
@@ -0,0 +1,59 @@
+const breakLoop = require('./breakLoop')
+
+// for async generators
+export default function asyncEachOfLimit(generator, limit, iteratee, callback) {
+    let done = false
+    let canceled = false
+    let awaiting = false
+    let running = 0
+    let idx = 0
+
+    function replenish() {
+        //console.log('replenish')
+        if (running >= limit || awaiting) return
+        //console.log('replenish awaiting')
+        awaiting = true
+        generator.next().then(({value, done: iterDone}) => {
+            //console.log('got value', value)
+            awaiting = false
+            if (iterDone) {
+                done = true;
+                if (running <= 0) {
+                    callback(null)
+                }
+                return;
+            }
+            running++
+            iteratee(value, idx, iterateeCallback)
+            idx++
+            replenish()
+        }).catch(handleError)
+    }
+
+    function iterateeCallback(err, result) {
+        //console.log('iterateeCallback')
+        if (canceled) return
+        running -= 1;
+        if (err) return handleError(err)
+
+        if (err === false) {
+            done = true;
+            canceled = true;
+        }
+
+        if (result === breakLoop || (done && running <= 0)) {
+            done = true;
+            //console.log('done')
+            return callback(null);
+        }
+        replenish()
```

Comment:
I think you have to avoid calling `replenish` when `done === true` or have replenish check if `done === true`.

Lets say hypothetically we have a limit of 2 and the final 2 `iteratees` are running. When one of these iteratees resolves, `running = 1` and `done = true` but we'll still replenish and call the `generator.next().then()`. This has the possibility of entering a race condition, if the other iteratee resolves in the time between the `then` callback is called, you can actually end up calling `callback` twice